### ISSUE TITLE
Experiments are not fully-reproducable

### ISSUE THREAD

**Issue by Vectorrent** [0-5788719150923125/praxis#9]
While we have made some efforts to make experiments reproducible (via seeding), there are places where this has not been addressed yet. Among them:

- datasets, in particular the weighting strategy
- layer shuffling, just uses `random.shuffle()` currently

There are probably other places, too. It would be nice to write tests to validate full reproducibility.

**Comment by Vectorrent** [comment]
From [Reddit](https://www.reddit.com/r/pytorch/comments/fo23lt/cant_achive_reproducability_determinism_in/)

_If you really want complete reproducibility you have to disable multithreading, this is not possible on GPU._

_Why?_

_Because with float32, due to rounding (a + b) + c is different from a + (b + c). I.e. floating point addition is not associative._

_When you reduce_sum or do a softmax or any loss function that reduces in parallel a tensor to a single value, the order of which sum is done first by what thread is completely non-deterministic._

I got us as far as could be expected, without disabling multithreading. At the end of the day, even setting the Pytorch `determinism=True` flag does not lead to completely deterministic results; if you inspect losses for two simple models, the precise hills and valleys in their losses will quickly diverge. There is no way to avoid this, so far as I can tell.

Nevertheless, our model is *as deterministic as possible*, with one exception. DataLoaders are completely deterministic, and will always retrieve, prepare, batch and dispatch data in the same order. This is an important milestone.

However, in the end, we decided to set `deterministic=False`, because it doesn't work with `benchmark=True` - and disabling benchmarking results in about 10% performance degradation (in speed). I don't find such a degradation acceptable, when determinism still cannot be guaranteed. And regardless, the deterministic results are still comparable, with or without the `deterministic` flag.

I'm not going to fight with this any longer. We've taken it as far as we can for now. Experiments are "mostly deterministic" - and that's good enough for me.
